cleverhans/attacks_tf.py

```diff
@@ -1320,3 +1320,155 @@ def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
+            grad = self.sess.run(self.grad, feed_dict={
+                                 self.x: adv_x.reshape(oimgs.shape),
+                                 self.targeted_label: targets,
+                                 self.ori_img: oimgs,
+                                 self.const: CONST})
+            return loss, grad.flatten().astype(float)
+
+        # begin the main part for the attack
+        from scipy.optimize import fmin_l_bfgs_b
+        oimgs = np.clip(x_val, self.clip_min, self.clip_max)
+        CONST = np.ones(self.batch_size) * self.initial_const
+
+        # set the lower and upper bounds accordingly
+        lower_bound = np.zeros(self.batch_size)
+        upper_bound = np.ones(self.batch_size) * 1e10
+
+        # set the box constraints for the optimization function
+        clip_min = self.clip_min*np.ones(oimgs.shape[:])
+        clip_max = self.clip_max*np.ones(oimgs.shape[:])
+        clip_bound = list(zip(clip_min.flatten(), clip_max.flatten()))
+
+        # placeholders for the best l2 and instance attack found so far
+        o_bestl2 = [1e10] * self.batch_size
+        o_bestattack = np.copy(oimgs)
+
+        for outer_step in range(self.binary_search_steps):
+            _logger.debug(("  Binary search step {} of {}")
+                          .format(outer_step, self.binary_search_steps))
+
+            # The last iteration (if we run many steps) repeat the search once.
+            if self.repeat and outer_step == self.binary_search_steps - 1:
+                CONST = upper_bound
+
+            # optimization function
+            adv_x, f, _ = fmin_l_bfgs_b(lbfgs_objective,
+                                        oimgs.flatten().astype(float),
+                                        args=(self, targets, oimgs, CONST),
+                                        bounds=clip_bound,
+                                        maxiter=self.max_iterations,
+                                        iprint=0)
+
+            adv_x = adv_x.reshape(oimgs.shape)
+            assert np.amax(adv_x) <= self.clip_max and \
+                np.amin(adv_x) >= self.clip_min, \
+                'fmin_l_bfgs_b returns are invalid'
+
+            # adjust the best result found so far
+            preds = np.atleast_1d(utils_tf.model_argmax(self.sess, self.x,
+                                                        self.model_preds,
+                                                        adv_x))
+            _logger.debug("predicted labels are {}".format(preds))
+
+            l2s = np.zeros(self.batch_size)
+            for i in range(self.batch_size):
+                l2s[i] = np.sum(np.square(adv_x[i] - oimgs[i]))
+
+            for e, (l2, pred, ii) in enumerate(zip(l2s, preds, adv_x)):
+                if l2 < o_bestl2[e] and pred == np.argmax(targets[e]):
+                    o_bestl2[e] = l2
+                    o_bestattack[e] = ii
+
+            # adjust the constant as needed
```

Comment:
can you describe the expected behavior more explicitly here?

Reply:
I think the descriptions in line 1453 and line 1458-1459 are enough to explain how search is conducted?